### PR TITLE
chore: prepare version 7

### LINKED ISSUE #1973: Make debug log easier to structurally ingest into a logging system like Logstash/Elasticsearch
I'd really like to ingest the debug log into my ELK (Elasticsearch, Logstash & Kibana) stack so that I can more easily analyze it for anomalies and graph how often issues occur in my Z-Wave mesh. For this to be possible, the information in the logs must be separated into more atomic units so that it can be parsed in a simple way.

For the logs to be easy to ingest with a log parser they must be **structured** in a way that makes it possible to _know where one value starts and ends and the next value begins_. In the current log this is mostly implemented by using indenting with space. It is also _vital to know what the value represents_. This is currently implemented with fixed line character offsets, and the message part always start at a certain offset of the line. Unfortunately this only means the message key begins there, the message **value** is still just opaque text with no inherent structure. Some long sentences are also split over multiple lines, although they technically don't need to be. This seems to have been done to keep the log max line length to a fixed number. I'm assuming this is to accommodate easier human consumption on a screen with limited width.

Introducing a structured logging format would still make it possible to read for humans and also ingestable by automatic tools. It would also make it easier to search using `grep(1)` and filter with tools like `sed(1)`. My suggestion is to use the [logfmt](https://www.brandur.org/logfmt) format championed by Heroku. There is even an [npm package available](https://www.npmjs.com/package/logfmt) for it.

Here are some examples of how log data from my own log could be represented using logfmt fully structured:

```
20:23:02.101 SERIAL » 0x01030015e9                                                         (5 bytes)
ts=2021-03-01T20:23:02.101Z c=SERIAL dir=tx msg="Transmit payload" payload="0x01030015e9" length=5

20:23:02.101 DRIVER » [REQ] [GetControllerVersion]
ts=2021-03-01T20:23:02.101Z c=DRIVER dir=tx msg="GetControllerVersion" cmd="REQ"

20:23:02.102 SERIAL « [ACK]                                                                   (0x06)
ts=2021-03-01T20:23:02.102Z c=SERIAL dir=rx msg="Payload received" payload="0x06" length=1 cmd="ACK"

20:23:02.103 SERIAL « 0x011001155a2d5761766520332e3935000199                              (18 bytes)
ts=2021-03-01T20:23:02.103Z c=SERIAL dir=rx msg="Payload received" payload="0x011001155a2d5761766520332e3935000199" length=18

20:23:02.104 DRIVER « [RES] [GetControllerVersion]
                        payload: 0x5a2d5761766520332e39350001
ts=2021-03-01T20:23:02.104Z c=DRIVER dir=rx msg="GetControllerVersion" payload="0x5a2d5761766520332e39350001" cmd="RES"

20:23:02.105 CNTRLR   received version info:
                        controller type: Static Controller
                        library version: Z-Wave 3.95
ts=2021-03-01T20:23:02.105Z c=CNTRLR msg="received version info" controller_type="Static Controller" library_version="Z-Wave 3.95"

20:23:02.122 CNTRLR   received API capabilities:
                        serial API version:  1.0
                        manufacturer ID:     0x86
                        product type:        0x01
                        product ID:          0x5a
                        supported functions:
                        · GetSerialApiInitData (0x02)
                        · FUNC_ID_SERIAL_API_APPL_NODE_INFORMATION (0x03)
                        · ApplicationCommand (0x04)
                        · GetControllerCapabilities (0x05)
ts=2021-03-01T20:23:02.122Z c=CNTRLR msg="received API capabilities" serial_api_version=1.0 manufacturer_id=0x86 product_type=0x01 product_id=0x5a
ts=2021-03-01T20:23:02.122Z c=CNTRLR msg="received API capabilities" supported_function_id=0x02 supported_function_name=GetSerialApiInitData
ts=2021-03-01T20:23:02.122Z c=CNTRLR msg="received API capabilities" supported_function_id=0x03 supported_function_name=FUNC_ID_SERIAL_API_APPL_NODE_INFORMATION"
ts=2021-03-01T20:23:02.122Z c=CNTRLR msg="received API capabilities" supported_function_id=0x04 supported_function_name=ApplicationCommand
ts=2021-03-01T20:23:02.122Z c=CNTRLR msg="received API capabilities" supported_function_id=0x05 supported_function_name=GetControllerCapabilities

20:23:02.286 DRIVER   Cache file for homeId 0xdc364e03 found, attempting to restore the network from
                       cache...
ts=2021-03-01T20:23:02.286Z c=DRIVER msg="Cache file for home identifier found, attempting to restore the network from cache" home_id=0xdc364e03

20:23:02.275 CNTRLR   received node information:
                        controller role:            primary
                        controller is a SUC:        true
                        controller is a slave:      false
                        controller supports timers: false
                        nodes in the network:       1, 3, 15, 17, 18, 22, 23, 24, 26, 30, 31
ts=2021-03-01T20:23:02.275Z c=CNTRLR msg="received node information" controller_role=primary controller_is_a_suc=true controller_is_a_slave=false controller_supports_timers=false nodes_in_network="1,3,15,17,18,22,23,24,26,30,31"

20:23:02.287 CNTRLR   [Node 001] trying to load device config
ts=2021-03-01T20:23:02.287Z c=CNTRLR node_id=1 msg="trying to load device config"

20:23:02.330 CNTRLR   [Node 001] [Manufacturer Specific] manufacturerId: metadata updated
20:23:02.330 CNTRLR   [Node 001] [~] [Manufacturer Specific] manufacturerId: 134 => 134
ts=2021-03-01T20:23:02.330Z c=CNTRLR node_id=1 msg="metadata updated" cc="Manufacturer Specific" key=manufacturerId value=134 old_value=134
```

If going full-blown with structured logging is too much to bite over in one go, you could start simple by encoding the existing message with the use of indent, but add some structure to make it easier to parse. Under you can see an example of this half-way approach shown on the same messages:

```
20:23:02.101 SERIAL » 0x01030015e9                                                         (5 bytes)
ts=2021-03-01T20:23:02.101Z c=SERIAL dir=tx msg="Transmit payload" payload="0x01030015e9" length=5

20:23:02.101 DRIVER » [REQ] [GetControllerVersion]
ts=2021-03-01T20:23:02.101Z c=DRIVER dir=tx msg="GetControllerVersion" cmd="REQ"

20:23:02.102 SERIAL « [ACK]                                                                   (0x06)
ts=2021-03-01T20:23:02.102Z c=SERIAL dir=rx msg="[ACK]" payload="0x06" length=1

20:23:02.103 SERIAL « 0x011001155a2d5761766520332e3935000199                              (18 bytes)
ts=2021-03-01T20:23:02.103Z c=SERIAL dir=rx msg="0x011001155a2d5761766520332e3935000199" length=18

20:23:02.104 DRIVER « [RES] [GetControllerVersion]
                        payload: 0x5a2d5761766520332e39350001
ts=2021-03-01T20:23:02.104Z c=DRIVER dir=rx part=1/2 msg="[RES] [GetControllerVersion]"
ts=2021-03-01T20:23:02.104Z c=DRIVER dir=rx part=2/2 msg="  payload: 0x5a2d5761766520332e39350001"

20:23:02.105 CNTRLR   received version info:
                        controller type: Static Controller
                        library version: Z-Wave 3.95
ts=2021-03-01T20:23:02.105Z c=CNTRLR part=1/3 msg="received version info:"
ts=2021-03-01T20:23:02.105Z c=CNTRLR part=2/3 msg="  controller type: Static Controller"
ts=2021-03-01T20:23:02.105Z c=CNTRLR part=3/3 msg="  library version: Z-Wave 3.95"

20:23:02.122 CNTRLR   received API capabilities:
                        serial API version:  1.0
                        manufacturer ID:     0x86
                        product type:        0x01
                        product ID:          0x5a
                        supported functions:
                        · GetSerialApiInitData (0x02)
                        · FUNC_ID_SERIAL_API_APPL_NODE_INFORMATION (0x03)
                        · ApplicationCommand (0x04)
                        · GetControllerCapabilities (0x05)
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=1/10 msg="received API capabilities:"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=2/10 msg="  serial API version:  1.0"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=3/10 msg="  manufacturer ID:     0x86"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=4/10 msg="  product type:        0x01"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=5/10 msg="  product ID:          0x5a"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=6/10 msg="  supported functions:"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=7/10 msg="  · GetSerialApiInitData (0x02)"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=8/10 msg="  · FUNC_ID_SERIAL_API_APPL_NODE_INFORMATION (0x03)"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=9/10 msg="  · ApplicationCommand (0x04)"
ts=2021-03-01T20:23:02.122Z c=CNTRLR part=10/10 msg="  · GetControllerCapabilities (0x05)"

20:23:02.286 DRIVER   Cache file for homeId 0xdc364e03 found, attempting to restore the network from
                       cache...
ts=2021-03-01T20:23:02.286Z c=DRIVER msg="Cache file for homeId 0xdc364e03 found found, attempting to restore the network from cache"

20:23:02.287 CNTRLR   [Node 001] trying to load device config
ts=2021-03-01T20:23:02.287Z c=CNTRLR node_id=1 msg="trying to load device config"

20:23:02.330 CNTRLR   [Node 001] [Manufacturer Specific] manufacturerId: metadata updated
20:23:02.330 CNTRLR   [Node 001] [~] [Manufacturer Specific] manufacturerId: 134 => 134
ts=2021-03-01T20:23:02.330Z c=CNTRLR node_id=1 part=1/2 msg="[Manufacturer Specific] manufacturerId: metadata updated"
ts=2021-03-01T20:23:02.330Z c=CNTRLR node_id=1 part=2/2 msg="[~] [Manufacturer Specific] manufacturerId: 134 => 134"
```

The lines that start with `ts=` are logfmt format. With this halfway approach you could most likely convert existing logging to use this new format with not too much effort, and then you could slowly move over to the fully structured format as time permits on individual pieces of code as you work on them.

I don't know what all the log information represents, so I've taken liberties with naming the keys in the logfmt syntax. The only one that is commonly referred to in the logfmt documentation is `msg`, and it should be kept named as is. I've also seen `ts` being used in lots of places to represents the timestamp, other then that it is up to you if you'd prefer to go with short 1, 2 or 3-character keys or fully named for what they represent. I think the logfmt format generally use underscores, but if everything is already in camelCase I don't see a reason why you can't use that. I tried to come up with ways to represent both key/value maps and arrays in the log syntax, as that can be a point of challenge when you're trying to dump complex structures to simple lines.

I hope this hasn't been too much of a rambling mess and that it makes sense to you. I hope you will consider it.

### PR DESCRIPTION
**Pull request by AlCalzone**
fixes: #1973

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,23 @@
 	Placeholder for next release:
 	## __WORK IN PROGRESS__
 -->
+## __WORK IN PROGRESS__ · _Summer is coming!_
+### Breaking changes · [Migration guide](https://zwave-js.github.io/node-zwave-js/#/getting-started/migrating-to-v7)
+* Renamed `controller.removeNodeFromAllAssocations` to `controller.removeNodeFromAllAssociations` to fix a typo
+* We've reworked/fixed the parsing of Node Information Frames (NIF) to match the specifications and changed node properties to make more sense
```
Comment (robertsLando) [inline CHANGELOG.md:9]: I also remember something about the `interview completed` event that isn't triggered anymore right?

Comment (marcus-j-davies) [reply CHANGELOG.md:9]: I'll start testing this very shortly.

Nice work 👍

Comment (AlCalzone) [reply CHANGELOG.md:9]: https://github.com/zwave-js/node-zwave-js/pull/2030/files?file-filters%5B%5D=.md#diff-0673db990a9385cee3ac29f1a4709dfeb8ad1cfe59fc9267ef4ad81b7de1ef13R5

File: docs/getting-started/migrating-to-v7.md [hunk 2]
```diff
@@ -4,10 +4,12 @@ In version 7.x we got rid of some old habits, leading to several breaking change
 
 ## No automatic query of all node values when restarting from cache
 
-This legacy behavior resulted in a lot of traffic and delays when the driver was restarted, often doing many unnecessary queries. To reduce the strain on battery devices and to keep the network as responsive as possible, we've opted not to do this anymore when a node was previously interviewed. This also means that the `"interview completed"` event is no longer emitted on restart.
+This legacy behavior resulted in a lot of traffic and sdelays when the driver was restarted, often doing many unnecessary queries. To reduce the strain on battery devices and to keep the network as responsive as possible, we've opted not to do this anymore when a node was previously interviewed. This also means that the `"interview completed"` event is only emitted after the initial interview and after manually-requested re-interviews, but not after each restart.
```
Comment (marcus-j-davies) [inline docs/getting-started/migrating-to-v7.md]: > This legacy behavior resulted in a lot of traffic and sdelays

s in delays - its only small, but let me have this moment 😄

Comment (AlCalzone) [reply docs/getting-started/migrating-to-v7.md]: that's a typo. Probably wanted to save the document :D

**Comment by marcus-j-davies** [comment]
I'm going to start testing out stuff (that affects me/node-red)
how do you want bugs reported? - v7 in the title?

**Comment by AlCalzone** [comment]
Sounds good (if you can isolate it to v7)

**Comment by marcus-j-davies** [comment]
@AlCalzone,

I tested the changes you implemented: (https://github.com/zwave-js/node-zwave-js/issues/2034)

And I was still getting issues (same error).

The assertion failure is on :
`validatePayload(eventDataLength === 16 || eventDataLength === 32);`

the `eventDataLength` will be reporting the size of the user consumable data (in this case anyway)
I'm not sure if that should/shouldn't include the padding? - i don't know enough about the spec.

But `eventDataLength` is > 0 but isn't 16 | 32 - it's the size of the readable sting.
which again - I don't know should actually include the padding.

if the length field, should include padding bytes & for the lib to remain compliant, can we do something here with a compat flag?
I'll be surprised, if the data length filed for this CC, is not a failure in spec compliance for others.

removing this assertion check, will work, but to remain compliant, it should stay, but having a compat flag somewhere for it?

EDIT.
I checked the length of the payload after  `eventDataLength`.
and it still didn't include padding - so its clearly a failure on spec.

**Pull request merged**